illumio_connector.py

```diff
@@ -152,6 +151,286 @@ def _handle_get_traffic_analysis(self, param):
+        action_result.add_data(result)
+        return action_result.get_status()
+
+    def _handle_provision_objects(self, param):
+        hrefs = param["hrefs"]
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        ret_val = self.connect_pce(action_result)
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        try:
+            hrefs_list = self.handle_comma_seperated_string(hrefs)
+
+            provisioned_virtual_service_obj = self._pce.provision_policy_changes(
+                change_description="Phantom object provisioning.",
+                hrefs=hrefs_list,
+            )
+
+        except IllumioException as e:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                "Encountered error provisioning object: {}".format(e),
+            )
+
+        result = self.convert_object_to_json(
+            provisioned_virtual_service_obj, action_result
+        )
+        action_result.add_data(result)
+        return action_result.set_status(
+            phantom.APP_SUCCESS, "Successfully provisioned object"
+        )
+
+    def _handle_get_ip_lists(self, param):
+        name = param["name"]
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        ip_list_data = None
+
+        ret_val = self.connect_pce(action_result)
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        try:
+            ip_list = self._pce.ip_lists.get(
+                policy_version="active", params={"name": name}
+            )
+            for data in ip_list:
+                if name == data.name:
+                    ip_list_data = data
+                    action_result.set_status(
+                        phantom.APP_SUCCESS, "Successfully fetched IP List"
+                    )
+                    break
+            if not ip_list_data:
+                ip_list_data = {}
+                action_result.set_status(
+                    phantom.APP_ERROR,
+                    "IP List with name '{}' does not exist".format(name),
+                )
+
+        except IllumioException as e:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                "Encountered error fetching IP List: {}".format(e),
+            )
+
+        result = self.convert_object_to_json(ip_list_data, action_result)
+        action_result.add_data(result)
+        return action_result.get_status()
+
+    def _handle_create_ruleset(self, param):
+        name = param["name"]
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        rule_set = None
+
+        ret_val = self.connect_pce(action_result)
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        try:
+            ruleset_list = self._pce.rule_sets.get(params={"name": name})
```

Comment:
This pattern (GET by name, then create) is clunky, and we'll need to repeat it for each create function. It will be less verbose to reverse this - try to create first, then if a 4xx error is returned GET the existing object - but I'll look at adding a `PolicyComputeEngine` function that handles this internally.

Reply:
Changed the code to create a ruleset first, check if "rule_set_name_in_use" is returned in the error message, and return the existing ruleset else return the one that is created right now. We checked for the "rule_set_name_in_use" message instead of the 406 error because we want to fetch the existing ruleset only in case it is present and 406 may occur in other cases as well.